src/ResourceManager/Common/Commands.ResourceManager.Common/RPRegistrationDelegatingHandler.cs

```diff
@@ -0,0 +1,103 @@
+        {
+            registeredProviders = new HashSet<string>();
+            this.writeVerbose = writeVerbose;
+            this.createClient = createClient;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            ResourceManagementClient = createClient();
+
+            HttpResponseMessage responseMessage = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            if (IsProviderNotRegistereError(responseMessage))
+            {
+                var providerName = GetProviderName(request.RequestUri);
+                if (!string.IsNullOrEmpty(providerName) && !registeredProviders.Contains(providerName))
+                {
+                    registeredProviders.Add(providerName);
+                    try
+                    {
+                        writeVerbose(string.Format("Attempting to register resource provider '{0}'", providerName));
+                        ResourceManagementClient.Providers.Register(providerName);
+                        Provider provider = null;
+                        do
+                        {
+                            provider = ResourceManagementClient.Providers.Get(providerName).Provider;
+                        } while (provider.RegistrationState != RegistrationState.Registered.ToString());
+                        writeVerbose(string.Format("Succeeded to register resource provider '{0}'", providerName));
+                    }
+                    catch
+                    {
+                        writeVerbose(string.Format("Failed to register resource provider '{0}'", providerName));
+                        // Ignore RP registration errors.
+                    }
+
+                    responseMessage = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            return responseMessage;
+        }
+
+        private bool IsProviderNotRegistereError(HttpResponseMessage responseMessage)
+        {
+            return responseMessage.StatusCode == System.Net.HttpStatusCode.Conflict &&
+                responseMessage.Content.ReadAsStringAsync().Result.Contains("registered to use namespace");
+        }
+
+        /// <summary>
+        /// Extract provider name from request uri such as
+        /// "https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group-name}/providers/{resource-provider-name}"
+        /// We analyze the uri's segments and check the index of 5 is "providers/" and return the next segment.    
+        /// </summary>
+        /// <param name="requestUri">request uri to extract provider out</param>
+        /// <returns>provider name, or null on unexpected format</returns>
+        private string GetProviderName(Uri requestUri)
+        {
+            return (requestUri.Segments.Length > 7 && requestUri.Segments[5].ToLower() == "providers/") ?
```

Comment:
isn't it segments[4]? 0 = subscriptions, 1- subscriptionId, 2 = resource groups, 3= resource group name, 4 = providers